linkml/generators/fhirgen.py

```diff
@@ -0,0 +1,393 @@
+        :param inferred: (classes only) show all induced slots as attributes
+        :return: yaml string
+        """
+        if not inferred:
+            return yaml_dumper.dumps(element)
+        else:
+            if not isinstance(element, ClassDefinition):
+                raise ValueError(
+                    f"Inferred only applicable for classes, not {element.name} {type(element)}"
+                )
+            # TODO: move this code to schemaview
+            c = deepcopy(element)
+            attrs = self.schemaview.class_induced_slots(c.name)
+            for a in attrs:
+                c.attributes[a.name] = a
+            c.slots = []
+            return yaml_dumper.dumps(c)
+
+    def all_enum_objects(self) -> Iterator[EnumDefinition]:
+        """
+        all enum objects in schema
+
+        Ensures rank is non-null
+        :return: iterator
+        """
+        elts = self.schemaview.all_enums().values()
+        _ensure_ranked(elts)
+        for e in elts:
+            yield e
+
+    def class_hierarchy_as_tuples(self) -> Iterator[Tuple[int, ClassDefinitionName]]:
+        """
+        Generate a hierarchical representation of all classes in the schema
+
+        This is represented as a list of tuples (depth: int, cls: ClassDefinitionName),
+        where the order is pre-order depth first traversal
+
+        This can then be used to draw a hierarchy within jinja in a number of ways; e.g
+
+        - markdown (by using indentation of " " * depth on a list)
+        - tables (by placing fake indentation e.g using underscores)
+
+        Simply iterate through all tuples, drawing each in the appropriate way
+
+        Note: By default all classes are ordered alphabetically
+
+        :return: tuples (depth: int, cls: ClassDefinitionName)
+        """
+        sv = self.schemaview
+        roots = sv.class_roots(mixins=False)
+
+        # by default the classes are sorted alphabetically
+        roots = sorted(roots, key=str.casefold, reverse=True)
+
+        # the stack holds tuples of depth-class that have still to be processed.
+        # we seed this with all root classes (which have depth 0)
+        # note the stack is processed from the last element first, ie. LIFO
+        stack = [(0, root) for root in roots]
+        # use iterative depth first traversal
+        while len(stack) > 0:
+            depth, class_name = stack.pop()
+            yield depth, class_name
+            children = sorted(
+                sv.class_children(class_name=class_name, mixins=False),
+                key=str.casefold,
+                reverse=True,
+            )
+            for child in children:
+                # depth first - place at end of stack (to be processed next)
+                stack.append((depth + 1, child))
+
+
+# todo: Create an output file or just print to stdout as it is now?
+# TODO: test is generating an error: generator.decorator: default=g.valid_formats[0],
+# @shared_arguments(FhirValueSetGenerator)
```

Comment:
Not sure why, but when I ran my test, I got an error error in the `Generator` super class, in the `decorator` method, on `default=g.valid_formats[0],` (`valid_formats` was empty)).

Reply:
@hrshdhgd Any idea on this? 
I copied the code from `docgen`. And its CLI has `@shared_arguments(DocGenerator)`, so I added `@shared_arguments(FhirValueSetGenerator)` to my code. But when I do this...

I get an error raised here: [`default=g.valid_formats[0]`](https://github.com/linkml/linkml/blob/main/linkml/utils/generator.py#L812)

because [`g.vaid_formats` is an empty list](https://github.com/linkml/linkml/blob/main/linkml/utils/generator.py#L36).

I wonder why this happens when I try to use `@shared_arguments()` for my new generator, but it works fine for docgen.

Reply:
docgen has `valid_formats = ["markdown", "rst", "html", "latex"]` on line 100. Is there an equivalent for `fhirgen` ?

Reply:
Almost all `*gen.py` has a value for `valid_formats` . I think declaring that in `fhirgen.py` may resolve your issue.

Reply:
Fixed! thanks so much